File: go/collatz-conjecture/collatz_conjecture.go
package collatzconjecture

import "errors"

func CollatzConjecture(n int) (int, error) {
	if n < 1 {
		return 0, errors.New("n must be positive")
	}
	var step int
	for n > 1 {
		step++
		if n%2 == 0 {
			n /= 2
		} else {
			n = 3*n + 1
		}
	}
	return step, nil
}

File: clojure/collatz-conjecture/src/collatz_conjecture.clj
(ns collatz-conjecture)

(defn step [n]
  (if (even? n)
    (/ n 2)
    (inc (* n 3))))

(defn collatz [num]
  (assert (pos? num) "Input must be positive")
  (loop [n num
         acc 0]
    (if (= n 1)
      acc
      (recur (step n) (inc acc)))))
